src/core/icc_colorspace.js

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  DataType,
+  initSync,
+  Intent,
+  qcms_convert_array,
+  qcms_convert_four,
+  qcms_convert_one,
+  qcms_convert_three,
+  qcms_drop_transformer,
+  qcms_transformer_from_memory,
+} from "../../external/qcms/qcms.js";
+import { shadow, warn } from "../shared/util.js";
+import { ColorSpace } from "./colorspace.js";
+import { QCMS } from "../../external/qcms/qcms_utils.js";
+
+class IccColorSpace extends ColorSpace {
+  #transformer;
+
+  #convertPixel;
+
+  static #useWasm = true;
+
+  static #wasmUrl = null;
+
+  static #finalizer = new FinalizationRegistry(transformer => {
+    qcms_drop_transformer(transformer);
+  });
+
+  constructor(iccProfile, numComps) {
+    if (!IccColorSpace.isUsable) {
+      throw new Error("No ICC color space support");
+    }
+
+    super("ICCBased", numComps);
+
+    let inType;
+    switch (numComps) {
+      case 1:
+        inType = DataType.Gray8;
+        this.#convertPixel = (src, srcOffset) =>
+          qcms_convert_one(this.#transformer, src[srcOffset] * 255);
+        break;
+      case 3:
+        inType = DataType.RGB8;
+        this.#convertPixel = (src, srcOffset) =>
+          qcms_convert_three(
+            this.#transformer,
+            src[srcOffset] * 255,
+            src[srcOffset + 1] * 255,
+            src[srcOffset + 2] * 255
+          );
+        break;
+      case 4:
+        inType = DataType.CMYK;
+        this.#convertPixel = (src, srcOffset) =>
+          qcms_convert_four(
+            this.#transformer,
+            src[srcOffset] * 255,
+            src[srcOffset + 1] * 255,
+            src[srcOffset + 2] * 255,
+            src[srcOffset + 3] * 255
+          );
+        break;
+      default:
+        throw new Error(`Unsupported number of components: ${numComps}`);
+    }
+    this.#transformer = qcms_transformer_from_memory(
+      iccProfile,
+      inType,
+      Intent.Perceptual
+    );
+    if (!this.#transformer) {
+      throw new Error("Failed to create ICC color space");
+    }
+    IccColorSpace.#finalizer.register(this, this.#transformer);
+  }
+
+  getRgbItem(src, srcOffset, dest, destOffset) {
+    QCMS._destBuffer = dest.subarray(destOffset, destOffset + 3);
+    this.#convertPixel(src, srcOffset);
+    QCMS._destBuffer = null;
+  }
+
+  getRgbBuffer(src, srcOffset, count, dest, destOffset, bits, alpha01) {
+    QCMS._destBuffer = dest.subarray(
+      destOffset,
+      destOffset + count * (3 + alpha01)
+    );
+    qcms_convert_array(
+      this.#transformer,
+      src.subarray(srcOffset, srcOffset + count * this.numComps)
+    );
+    QCMS._destBuffer = null;
+  }
+
+  getOutputLength(inputLength, alpha01) {
+    return (inputLength / this.numComps) * (3 + alpha01);
```

Comment:
Similar to other colorSpaces, let's ensure that this is always a whole number.
```suggestion
    return ((inputLength / this.numComps) * (3 + alpha01)) | 0;
```